Add support for noko fragments

chrismytton opened the pull request:
This allows an optional `noko` parameter to `ScrapedPage` constructors, which when given will override the `noko` scope for the current instance.

```ruby
MemberRow.new(response: response, noko: noko.css('.member-row'))
```

Then in the `MemberRow` `noko` would refer to the `.member-row` element, rather than the whole page, but it would still have access to the full response.

# Related issues

I think this should close #4, because then we can just use `ScrapedPage` subclasses everywhere and just pass an extra `noko` param

# Notes to merger

- [x] This needs to have it's base changed to `master` once #21 is merged.

chrismytton committed:
Add support for noko fragments

This allows an optional `noko` argument to `Scraped::HTML` constructors,
which when given will override the `noko` scope.

chrismytton commented:
Example of this being used to remove a custom `NokogiriDocument` class in the Mongolia scraper can be seen in https://github.com/everypolitician-scrapers/mongolia-khurai-wp-multiple-terms/pull/8.

chrismytton commented:
@tmtmtmtm What are your thoughts on this approach? If it seems like a reasonable approach then I'll write some documentation so we can get this PR merged.

tmtmtmtm commented:
Presumably this would only be on `Noko::HTML` (or conceivably `::XML`) after https://github.com/everypolitician/scraped/pull/29?

chrismytton commented:
> Presumably this would only be on Noko::HTML (or conceivably ::XML) after #29?

You mean `Scraped::HTML`? If so then yep, I will update this after that's merged :)

chrismytton commented:
@tmtmtmtm 👀

tmtmtmtm reviewed (changes requested):
Mostly 👍 , but having the two tests, which are conceptually largely identical, be set up in slightly different ways makes it a little harder to follow what's going on that it ideally should be. 

I think it would be good to harmonise those as much as possible (e.g. either both do the `let(:page) …`, or neither does).

chrismytton committed:
Make Scraped::HTML tests consistent with each other

This makes the two tests that are largely identical behave in the same
way.

chrismytton commented:
> I think it would be good to harmonise those as much as possible (e.g. either both do the let(:page) …, or neither does).

Ooh yep, good catch! I've made the first test use `let(:page)` as well in https://github.com/everypolitician/scraped/pull/23/commits/d84d6a3734f7327c712d3527da867719cdb25aa0.

tmtmtmtm reviewed (commented):
👍  — however (and sorry for not mentioning this first time around), I think we really need some docs for this too. I suspect the best example is going to be dealing with a table, where you want a MemberRow class, or equivalent, for each person.

chrismytton committed:
Add documentation for page fragments to README

chrismytton commented:
@tmtmtmtm I've added some docs in https://github.com/everypolitician/scraped/pull/23/commits/98cb4b065c83637bbc89af1633689ffc6f3d37e6. I'm not totally sure about the example I've used, it seems a bit contrived, but I think it's probably Good Enough for now.

tmtmtmtm reviewed (changes requested):
I fear that the extra level of indirection here might make this a little too confusing as an initial example, especially as `members_term_10_table` is quite an odd field.

Perhaps drop the Table class, and go for something like this?:

```ruby
class MemberRow < Scraped::HTML
  field :name do
    noko.css('td')[2].text
  end

  field :party do
    noko.css('td')[3].text
  end
end

class AllMembersPage < Scraped::HTML
  field :members do
    noko.css('table.members-list tr').map do |row|
      MemberRow.new(response: response, noko: row)
    end
  end
end
```

chrismytton committed:
Better example of fragments in the README

Rather than having an extra level of indirection with tables this uses a
simpler example of just a page and some member rows.

chrismytton commented:
> I fear that the extra level of indirection here might make this a little too confusing as an initial example, especially as members_term_10_table is quite an odd field.
> 
> Perhaps drop the Table class, and go for something like this?:

👍  Yep your version is much nicer! Pushed another commit to change it to that.

tmtmtmtm reviewed (approved):
👍
